### DISCUSSION TITLE
SpawnCommand.set_environment_variables vs os.getenv

### DISCUSSION THREAD

**Discussion by replicaJunction** [wezterm/wezterm#6132] [Q&A] [answered]
### What Operating System(s) are you running on?

Windows

### WezTerm version

20240203-110809-5046fc22

### Ask your question!

My goal is to create a few entries in the launcher menu. Each will spawn a shell in a different directory with a different Python virtual environment active (managed by the Poetry command), and I'd like each to have a different tab title so I can keep track of them better.

My idea for now is to use process-level environment variables to keep track of which process is which. However, it seems like os.getenv is not picking up on the variables set in my SpawnCommand config, even though the shell itself knows about them.

Right now, I'm using entries in the launch_menu that look like this:

```lua
{
  label = "Blocks Shell",
  args = { "pwsh", "-NoProfile", "-Command", "poetry", "shell" },
  cwd = "C:/Users/username/Documents/Code/blocks",
  set_environment_variables = {
    MY_WEZTERM_TAB_TITLE = "BLOCKS",
  },
},

{
  label = "Web Shell",
  args = { "pwsh", "-NoProfile", "-Command", "poetry", "shell" },
  cwd = "C:/Users/username/Documents/Code/web",
  set_environment_variables = {
    MY_WEZTERM_TAB_TITLE = "WEB",
  },
}
```

I also have a custom function set up for format_tab_title. There's a bunch of formatting shenanigans that are working as expected, but here's the relevant part of the function:

```lua
local full_title = os.getenv("MY_WEZTERM_TAB_TITLE") or tab.active_pane.title
local title_without_exe = string.gsub(full_title, "%.exe$", "")
local trimmed_title = wezterm.truncate_right(title_without_exe, max_width - 6)

-- ...

return {
  -- ...
  { Text = " " .. trimmed_title .. " " },
  -- ...
}
```

Here's the problem: the `os.getenv("MY_WEZTERM_TAB_TITLE")` does not seem to be picking up the environment variables set by the SpawnCommand.set_environment_variables config key. I've confirmed that the environment vars exist in the shell with the correct value.

For testing, I went into my Windows user-level environment variables and set the variable there - and the tab formatting worked. Every tab was titled with the test value I set.

Am I using the set_environment_variables key wrong? Is there a different way to get process-level environment variables instead of using `os.getenv()`? Is there a better way to accomplish my main goal?

**Comment by wez** [comment, marked as answer]
`set_environment_variables` sets environment variables for the child processes that are spawned in the default execution domain. It does not set them in the wezterm process, which is what `os.getenv` reads.

In general, changing the environment in a running program is fraught with peril, as it leads to race conditions in multi-threaded programs, so it is something to be avoided. Also: it is impossible for one program to change the environment of another running program; you can only change your own environment, or prepare the environment to be used for a child process that you spawn, once, at the time that you spawn it.

I'm not super clear on what your higher level goal is.

If you want to declare a "variable" (or rather, a constant, that doesn't change) centrally in your config, you can use the lua `local` syntax for that and just declare the variable in your lua code, with no need for `os.getenv`.

If you want to maintain a variable that updates in response to things happening in your config, you might consider looking at [wezterm.GLOBAL](https://wezfurlong.org/wezterm/config/lua/wezterm/GLOBAL.html).

I think that you might benefit from looking at *user-vars* here. https://wezfurlong.org/wezterm/recipes/passing-data.html has some information on passing state back from a program and into wezterm.   You might consider having your powershell startup stuff look at the `MY_WEZTERM_TAB_TITLE` environment variable and in turn have it emit a user var back to wezterm so that you can read that and update your tab title accordingly.

**Reply by replicaJunction** [reply]
That explains it. I didn't put together that `os.getenv` was still running in Wezterm's scope instead of running in the shell's scope. Thanks very much for helping me to understand that.

My higher level goal is for each custom entry in the launcher menu to have a different tab title. The paths and titles are fairly static, so it wouldn't be a problem to update my config if they changed or I needed to add more. Now that I think it through, I guess I could just set a static mapping somewhere that maps a path to a title; I just thought of the environment variable idea first. 😆 

> You might consider having your powershell startup stuff look at the MY_WEZTERM_TAB_TITLE environment variable and in turn have it emit a user var back to wezterm so that you can read that and update your tab title accordingly.

This makes sense, but is it possible to set the starting value for those user variables directly from the launch menu entry?
